zk_toolbox/crates/zk_inception/src/utils/ports.rs

```diff
@@ -0,0 +1,350 @@
+            if let Some(chain_config) = ecosystem_config.load_chain(Some(chain)) {
+                dirs.push(chain_config.configs.clone())
+            }
+        }
+        dirs.push(ecosystem_config.link_to_code);
+
+        let mut ecosystem_ports = EcosystemPorts::default();
+        for dir in dirs {
+            if dir.is_dir() {
+                Self::scan_yaml_files(shell, &dir, &mut ecosystem_ports)
+                    .context(format!("Failed to scan directory {:?}", dir))?;
+            }
+        }
+
+        Ok(ecosystem_ports)
+    }
+
+    /// Scans the given directory for YAML files in the immediate directory only (non-recursive).
+    /// Processes each YAML file found and updates the EcosystemPorts accordingly.
+    fn scan_yaml_files(
+        shell: &Shell,
+        dir: &Path,
+        ecosystem_ports: &mut EcosystemPorts,
+    ) -> Result<()> {
+        for path in shell.read_dir(dir)? {
+            if !path.is_file() {
+                continue;
+            }
+            if let Some(extension) = path.extension() {
+                if extension == "yaml" || extension == "yml" {
+                    Self::process_yaml_file(shell, &path, ecosystem_ports)
+                        .context(format!("Error processing YAML file {:?}", path))?;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn process_yaml_file(
+        shell: &Shell,
+        file_path: &Path,
+        ecosystem_ports: &mut EcosystemPorts,
+    ) -> Result<()> {
+        let contents = shell.read_file(file_path)?;
+        let value: Value = serde_yaml::from_str(&contents)?;
+        Self::traverse_yaml(&value, "", file_path, ecosystem_ports);
+        Ok(())
+    }
+
+    fn traverse_yaml(
+        value: &Value,
+        path: &str,
+        file_path: &Path,
+        ecosystem_ports: &mut EcosystemPorts,
+    ) {
+        match value {
+            Value::Mapping(map) => {
+                for (key, val) in map {
+                    let new_path = if path.is_empty() {
+                        key.as_str().unwrap_or_default().to_string()
+                    } else {
+                        format!("{}:{}", path, key.as_str().unwrap_or_default())
+                    };
+
+                    if key.as_str() == Some("ports") {
+                        if let Value::Sequence(ports) = val {
+                            for port_entry in ports {
+                                if let Some(port_str) = port_entry.as_str() {
+                                    let parts: Vec<&str> = port_str.split(':').collect();
+
+                                    // Handle 3-part (IP:host:container) or 2-part (host:container) port mappings
+                                    let host_port_str = match parts.len() {
+                                        3 => parts.get(1),
+                                        2 => parts.first(),
+                                        _ => None,
+                                    };
```

Comment:
Just going to highlight cases where this wouldn't work and we can discuss whether it's too crazy to support:
- '[::1]:3010:3010' (IPV6)
- '${HOST_PORT:-3010}:3010' (env configurable, although nothing we can do about it)
- 3010-3014:3010-3014 (port ranges)

I suggest just doing `if parts.len() > 1` then take second to the last part.